exchange/bitswap/bitswap.go

```diff
@@ -317,8 +325,11 @@ func (bs *Bitswap) HasBlock(blk blocks.Block) error {
 	// it now as it requires more thought and isnt causing immediate problems.
 	bs.notifications.Publish(blk)
 
-	for _, s := range bs.SessionsForBlock(blk.Cid()) {
-		s.receiveBlockFrom("", blk)
+	k := blk.Cid()
+	ks := []*cid.Cid{k}
+	for _, s := range bs.SessionsForBlock(k) {
+		s.receiveBlockFrom(from, blk)
```

Comment:
I merged two functions. `from` will be "" when the user calls `HasBlock` but will be the peer from which we received the block when called from within bitswap.

Basically, we had two functions, HasBlock and a go routine inside ReceiveMessage where:

1. The second function would:
2. Notify all interested sessions that we had received the block.
3. Remove the block from sessoins' wantlists.
4. Call the first function which would:
5. Put the block in the blockstore.
6. Publish a notification on `bs.notifications`.
7. Notify the session that we received the block (again) but claiming to have received it from "". This was to prevent a race condition where we'd notify the session that we received the block before putting it in the blockstore and then someone would try to get the block from the blockstore, not find it, and ask bitswap for it. I believe we should also have removed this block from the session but I'm not sure why we didn't (the new code does).
8. Record the block in the bitswap engine.
9. Reprovide the block.

I got rid of the first function, moved step 3 into step 7, and made it notify the session with the correct peer in step 7 if we received the block from the network.

---

The bug was that race condition in step 7. The fix works iff you keep the same bitswap instance but, if you replace the bitswap instance with a new one, the caller will not find the block in the blockstore and ask the new bitswap instance for the block (which doesn't know about it).